src/schema/__tests__/FroidSchema.test.ts

```diff
@@ -0,0 +1,1690 @@
+        contractTags: ['storefront', 'internal'],
+      });
+
+      expect(actual).toEqual(
+        // prettier-ignore
+        gql`
+        extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+        scalar UsedCustomScalar1
+
+        scalar UsedCustomScalar2
+
+        enum UsedEnum {
+          VALUE_ONE
+          VALUE_TWO @inaccessible
+          VALUE_THREE
+        }
+
+        type Query {
+          node(id: ID!): Node @tag(name: "internal") @tag(name: "storefront")
+        }
+
+        interface Node @tag(name: "internal") @tag(name: "storefront") {
+          id: ID!
+        }
+
+        type User implements Node @key(fields: "userId customField1 customField2 customEnum1 customEnum2") {
+          id: ID!
+          userId: String!
+          customField1: UsedCustomScalar1
+          customField2: [UsedCustomScalar2!]!
+          customEnum1: UsedEnum
+          customEnum2: [UsedEnum!]!
+        }
+
+        type Todo implements Node @key(fields: "todoId customField") {
+          id: ID!
+          todoId: Int!
+          customField: UsedCustomScalar1
+        }
+      `
+      );
+    });
+
+    it('tags are identified from field arguments', () => {
+      const urlSchema = gql`
+        type TypeA @key(fields: "selections { selectionId }") {
+          selections: [TypeB!] @inaccessible
+          fieldWithArgument(argument: Int @tag(name: "storefront")): Boolean
+        }
+
+        type TypeB @key(fields: "selectionId", resolvable: false) {
+          selectionId: String!
+        }
+      `;
+
+      const altSchema = gql`
+        type TypeB @key(fields: "selectionId") {
+          selectionId: String! @tag(name: "storefront")
+        }
+      `;
+
+      const subgraphs = new Map();
+      subgraphs.set('url-subgraph', urlSchema);
+      subgraphs.set('alt-subgraph', altSchema);
+
+      const actual = generateSchema({
+        subgraphs,
+        froidSubgraphName: 'relay-subgraph',
+        contractTags: ['storefront', 'internal'],
+      });
+
+      expect(actual).toEqual(
+        // prettier-ignore
+        gql`
+        extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+        type Query {
+          node(id: ID!): Node @tag(name: "internal") @tag(name: "storefront")
+        }
+
+        interface Node @tag(name: "internal") @tag(name: "storefront") {
+          id: ID!
+        }
+
+        type TypeA implements Node @key(fields: "selections { __typename selectionId }") {
+          id: ID! @tag(name: "storefront")
+          selections: [TypeB!]
+        }
+
+        type TypeB implements Node @key(fields: "selectionId") {
+          id: ID! @tag(name: "storefront")
+          selectionId: String!
+        }
+      `
+      );
+    });
+  });
+
+  describe('when generating schema for complex keys', () => {
+    it('uses a custom key sorter to prefer complex keys', () => {
```

Comment:
?

Reply:
Yes. I want the flexibility for applying custom logic when selecting the key without it being purely ordinal.